lib/app/endpoint/terminology.rb

```diff
@@ -0,0 +1,224 @@
+                     params = [
+                       FHIR::Parameters::Parameter.new(name: 'result', valueBoolean: false),
+                       FHIR::Parameters::Parameter.new(name: 'cause', valueString: 'invalid'),
+                       FHIR::Parameters::Parameter.new(name: 'message', valueString: message)
+                     ]
+                     FHIR::Parameters.new(parameter: params)
+                   end
+          respond_with_type(retval, request.accept, 200)
+        else
+          issue = FHIR::OperationOutcome::Issue.new(severity: 'error', code: 'required', details: { text: 'The specified code system is not known by the terminology server' })
+          logger.warn "Need code system #{coding['system']}"
+          return respond_with_type(FHIR::OperationOutcome.new(issue: issue), request.accept, 400)
+        end
+      end
+
+      private
+
+      def validate_code(valueset, coding, _display)
+        valueset.contains_code?(coding)
+        # TODO: Figure out how to validate displays, seeing as we don't have access to displays in the valueset object
+      end
+
+      def get_valueset(id_param, url_param)
+        # if this param is present, the operation was called on a particular ValueSet instance
+        if id_param && !id_param.empty?
+          begin
+            valueset = Inferno::Terminology.get_valueset_by_id(params[:id_param])
+          rescue Inferno::Terminology::UnknownValueSetException
+            error_code = { code: 'MSG_NO_MATCH', display: "No ValueSet found matching the id '#{pid_param}''" }
+            logger.warn "Need Valueset #{id_param}"
+            issue = FHIR::OperationOutcome::Issue.new(severity: 'error', code: 'value', details: { coding: error_code })
+            return FHIR::OperationOutcome.new(issue: issue)
+          end
+        elsif url_param && !url_param.empty?
+          begin
+            valueset = Inferno::Terminology.get_valueset(url_param)
+          rescue Inferno::Terminology::UnknownValueSetException
+            error_code = { code: 'MSG_NO_MATCH', display: "No ValueSet found matching the URL '#{url_param}''" }
+            logger.warn "Need Valueset #{url_param}"
+            issue = FHIR::OperationOutcome::Issue.new(severity: 'error', code: 'value', details: { coding: error_code })
+            return FHIR::OperationOutcome.new(issue: issue)
+          end
+        elsif params[:context]
+          # TODO: implement me
+        else
+          issue_text = 'No parameter that identifies the terminology/Valueset to validate against was supplied'
+          issue = FHIR::OperationOutcome::Issue.new(severity: 'error', code: 'required', details: { text: issue_text })
+          return FHIR::OperationOutcome.new(issue: issue)
+        end
+        valueset.url
+      end
+
+      def respond_with_type(resource, accept, status = 200)
+        accept.each do |type|
```

Comment:
True. It's complicated because `accept` is an array, but we could just take the first item in the array (or search the array for either of the return types we care about before returning the default type).